### ISSUE TITLE
Implement POSIX file locking

### ISSUE THREAD

**Issue by andreasbaumann** [andreasbaumann/pgfuse#1]
For example using git on a pgfuse mounted directory fails due to lack of locking.

**Comment by andreasbaumann** [comment]
another candidate for problems ist O_CREAT and O_EXCL which are currently not handled..

**Comment by andreasbaumann** [comment]
Long time ago, so I forgot about FUSE completely. :-)
lock doesn't have to be implemented in the 'lock' callback but for network or special filesystems. Otherwise local locking inside FUSE should do all the work.

We have to treat 'open' flags correctly and then see if we still get those errors.

**Comment by andreasbaumann** [comment]
Aha, suspicious:

<pre>
May 06 20:54:29 eurohp1 pgfuse[16940]: Remove file '/.git/config.lock' on 'test/', thread #3071105536
May 06 20:54:29 eurohp1 postgres[10868]: WARNING:  there is already a transaction in progress
</pre>

**Comment by andreasbaumann** [comment]
Fixed bugs in transactions, but git behaves the same..

**Comment by andreasbaumann** [comment]
Checking the locking code in git and trying to produce a test case.
It's now definitely something with locking..

**Comment by andreasbaumann** [comment]
lockfile.c in git:

<pre>
int commit_lock_file_to(struct lock_file *lk, const char *path)
{
...
        if (rename(lk->filename.buf, path)) {
...
</pre>

This rename fails with:

<pre>
rename("/media/sd/abaumann/projects/pgfuse/tests/mnt/.git/config.lock", "/media/sd/abaumann/projects/pgfuse/tests/mnt/.git/config") = -1 EEXIST (File exists)
unlink("/media/sd/abaumann/projects/pgfuse/tests/mnt/.git/config.lock") = 0
write(2, "error: could not commit config f"..., 93error: could not commit config file /media/sd/abaumann/projects/pgfuse/tests/mnt/.git/config
) = 93
</pre>

So the problem is that a rename to an existing file should overwrite the exisiting file and not result
in EEXIST..
</pre>

**Comment by andreasbaumann** [comment]
The rename problem is solved, so a 'git init' works now.

Tried 'git add' and 'git commit', but after this I get:

<pre>
shell> git status
error: bad index file sha1 signature
fatal: index file corrupt
shell> git diff
error: bad index file sha1 signature
fatal: index file corrupt
</pre>

**Comment by andreasbaumann** [comment]
git add shows:

<pre>
add:link(".git/objects/d9/tmp_obj_eTkITa", ".git/objects/d9/b401251bb36c51ca5c56c2ffc8a24a78ff20ae") = -1 ENOSYS (Function not implemented)
</pre>

This can't be good. :-)

So we have to implement and test symlinks correctly (or at all)..

**Comment by andreasbaumann** [comment]
Quite some times are not set correctly, see separate bug #4.

**Comment by nicferrier** [comment]
I'm looking at this too now. Maybe I can get it working. Did you ever find a definitive reason why this won't work?

**Comment by andreasbaumann** [comment]
Yeah. It boils down to some bugs and not having implemented hard-links.. :-)
..and more to come.

I'm curious. What is the use case for running git on a FUSE filesystem?

**Comment by nicferrier** [comment]
Well, I'm thinking that it would be fun to write a github thing (I work in a large enterprise) with decent wiki and ticket integration and use postgres for the storage for the whole thing. That would make it scalable (it's easy to scale pg after all, with lots of peers). I could just store the git for such a thing in files... but it would be interesting if it was pg.

What I just discovered was that it only seems to be working copies that are problematic. It seems I can store bare git repos with pgfuse. Which actually is all I need. :-)

**Comment by andreasbaumann** [comment]
Ah, this is good news. :-)
